lib/spack/spack/spec.py

```diff
@@ -560,43 +560,47 @@ def __repr__(self):
 
 @key_ordering
 class DependencySpec(object):
-    """Dependencies can be one (or more) of several types:
+    """DependencySpecs connect two nodes in the DAG, and contain deptypes.
+
+    Dependencies can be one (or more) of several types:
 
     - build: needs to be in the PATH at build time.
     - link: is linked to and added to compiler flags.
     - run: needs to be in the PATH for the package to run.
 
     Fields:
-    - spec: the spack.spec.Spec description of a dependency.
-    - deptypes: strings representing the type of dependency this is.
+    - spec: Spec depended on by parent.
+    - parent: Spec that depends on `spec`.
+    - deptypes: list of strings, representing dependency relationships.
     """
 
-    def __init__(self, spec, deptypes, default_deptypes=False):
+    def __init__(self, parent, spec, deptypes):
+        self.parent = parent
         self.spec = spec
-        self.deptypes = deptypes
-        self.default_deptypes = default_deptypes
+        self.deptypes = tuple(sorted(set(deptypes)))
 
     def update_deptypes(self, deptypes):
-        if self.default_deptypes:
-            self.deptypes = deptypes
-            self.default_deptypes = False
-            return True
-        return False
-
-    def _cmp_key(self):
-        return self.spec
+        deptypes = tuple(sorted(set(deptypes)))
+        changed = self.deptypes != deptypes
+        self.deptypes = deptypes
+        return changed
 
     def copy(self):
```

Comment:
Another side question: is it worth to implementing `__copy__` and `__deepcopy__` instead of using this method? If we conform to the [copy](https://docs.python.org/2/library/copy.html) module protocol, then we gain the possibility to seamlessly copy e.g. list of Specs or dicts of Specs, etc.